packages/agent/src/oidc.ts

```diff
@@ -0,0 +1,703 @@
+    { name: 'HKDF' },
+    false,
+    ['deriveBits']
+  );
+  const derivedKeyBits = await crypto.subtle.deriveBits(
+    {
+      name : 'HKDF',
+      hash : 'SHA-256',
+      info : new Uint8Array(),
+      salt : new Uint8Array(),
+    },
+    derivedKey,
+    256
+  );
+  const sharedEncryptionKey = new Uint8Array(derivedKeyBits);
+  return sharedEncryptionKey;
+}
+
+function encryptAuthResponse({
+  jwt,
+  encryptionKey,
+  providerDidKid,
+  randomPin,
+}: {
+  jwt: string;
+  encryptionKey: Uint8Array;
+  providerDidKid: string;
+  randomPin: string;
+}) {
+  const protectedHeader = {
+    alg : 'dir',
+    cty : 'JWT',
+    enc : 'XC20P',
+    typ : 'JWT',
+    kid : providerDidKid,
+  };
+  const nonce = CryptoUtils.randomBytes(24);
+  const additionalData = Convert.object({
+    ...protectedHeader,
+    pin: randomPin,
+  }).toUint8Array();
+
+  const jwtBytes = Convert.string(jwt).toUint8Array();
+  const chacha = xchacha20poly1305(encryptionKey, nonce, additionalData);
+  const ciphertextAndTag = chacha.encrypt(jwtBytes);
+
+  /** The cipher output concatenates the encrypted data and tag
+   * so we need to extract the values for use in the JWE. */
+  const ciphertext = ciphertextAndTag.subarray(0, -16);
+  const authenticationTag = ciphertextAndTag.subarray(-16);
+
+  const compactJwe = [
+    Convert.object(protectedHeader).toBase64Url(),
+    '', // Empty string since there is no wrapped key.
+    Convert.uint8Array(nonce).toBase64Url(),
+    Convert.uint8Array(ciphertext).toBase64Url(),
+    Convert.uint8Array(authenticationTag).toBase64Url(),
+  ].join('.');
+
+  return compactJwe;
+}
+
+export async function createPermissionGrants(
+  selectedDids: string[],
+  delegateDid: BearerDid,
+  dwn: AgentDwnApi
+) {
+  const grantPromises = selectedDids.map(async (did) => {
+    // TODO: Replace with real permission request
+    const permissionRequestData = {
+      description : 'The app is asking to Records Write to http://profile-protocol.xyz',
+      scope       : {
+        interface : DwnInterfaceName.Records,
+        method    : DwnMethodName.Write,
+        protocol  : 'http://profile-protocol.xyz',
+      },
+    };
+
+    // TODO: Confirm this
+    const message = await dwn.processRequest({
+      author        : did,
+      target        : did,
+      messageType   : DwnInterface.RecordsWrite,
+      messageParams : {
+        recipient    : delegateDid.uri,
+        protocolPath : 'grant',
+        protocol     : ' https://tbd.website/dwn/permissions',
+        dataFormat   : 'application/json',
+        data         : Convert.object(permissionRequestData).toUint8Array(),
+      },
+      // todo: is it data or datastream?
+      // dataStream: await Convert.object(permissionRequestData).toBlobAsync(),
+    });
+
+    return message;
+  });
+
+  const permissionGrants = (await Promise.all(grantPromises)).flat();
+
+  return permissionGrants;
+}
+
+async function submitAuthResponse(
+  selectedDids: string[],
```

Comment:
Sanity: It seems like we are only ever using one DID, is there a reason for this to be an Array vs just `selectedDid: string`?

Reply:
Good call will fix

Reply:
fixed